contracts/competition/Competition.sol

```diff
@@ -0,0 +1,138 @@
+        address[] members;
+        address[] joinRequests;
+    }
+
+    address[] private leaders;
+    mapping(address => Team) private teams;
+    mapping(string => bool) private teamNames;
+    mapping(address => address) private membersToTeam;
+    mapping(address => mapping(address => bool)) private requests;
+
+    modifier registrationIsOpen()
+    {
+        require(block.timestamp >= registrationStart, "Registration is not opened yet.");
+        require(block.timestamp < registrationEnd, "Registration is closed.");
+        _;
+    }
+
+    modifier isNotLeader()
+    {
+        require(teams[msg.sender].leader == address(0), "Team leaders are not allowed.");
+        _;
+    }
+
+    modifier isNotMember()
+    {
+        require(membersToTeam[msg.sender] == address(0), "Team members are not allowed.");
+        _;
+    }
+
+    constructor (
+        uint _start,
+        uint _end,
+        uint _registrationStart,
+        uint _registrationEnd,
+        IReferralStorage _referralStorage
+    ) public
+    {
+        start = _start;
+        end = _end;
+        registrationStart = _registrationStart;
+        registrationEnd = _registrationEnd;
+        referralStorage = _referralStorage;
+    }
+
+    function registerTeam (string calldata name, bytes32 referral) external registrationIsOpen isNotLeader
+    {
+        require(referralStorage.codeOwners(referral) != address(0), "Referral code does not exist.");
+        require(teamNames[name] == false, "Team name already registered.");
+
+        Team storage team;
+
+        team.leader = msg.sender;
+        team.name = name;
+        team.referral = referral;
+        team.members.push(msg.sender);
+
+        teams[msg.sender] = team;
+        leaders.push(msg.sender);
+        teamNames[name] = true;
+    }
+
+    function createJoinRequest (address leaderAddress) external registrationIsOpen isNotLeader isNotMember
+    {
+        require(membersToTeam[msg.sender] == address(0), "You can't join multiple teams.");
+        require(teams[msg.sender].leader != address(0), "The team does not exist.");
+        require(requests[leaderAddress][msg.sender] == false, "You already applied for this team.");
+
+        teams[leaderAddress].joinRequests.push(msg.sender);
+        requests[leaderAddress][msg.sender] = true;
+    }
+
+    function approveJoinRequest (address memberAddress) external registrationIsOpen isNotMember
+    {
+        require(requests[msg.sender][memberAddress] == false, "This member did not apply.");
+        require(membersToTeam[memberAddress] == address(0), "This member already joined a team.");
+
+        referralStorage.setTraderReferralCode(memberAddress, teams[msg.sender].referral);
+        teams[msg.sender].members.push(msg.sender);
+        membersToTeam[memberAddress] = msg.sender;
+    }
+
+    function getLeaders() external view returns (address[] memory)
+    {
+        address[] memory res;
+
+        for (uint i = 0; i < leaders.length; i++) {
```

Comment:
theoretically if there are too many leaders that function may run out of gas. in this specific case this is unlikely i believe, however it will be good to add start/end args to this func. 

btw will it we used by UI? to display list of all teams?

Reply:
Yup, same as above (used for leaderboard but will probably be removed if we use the graph)

Reply:
okay, in case we will have that method in contract would be good to add start/end args